src/main/java/io/anserini/collection/ParquetDenseVectorCollection.java

```diff
@@ -152,19 +153,24 @@ protected synchronized void readNext() throws IOException, NoSuchElementExceptio
         throw new NoSuchElementException("End of file reached");
       }
       
-    // Read each record from the Parquet file
-      // Extract the docid (String) from the record
       String docid = record.getString("docid", 0);
       ids.add(docid);
 
-      // Extract the vector (double[]) from the record
-      Group vectorGroup = record.getGroup("vector", 0); // Access the 'vector' field
-      int vectorSize = vectorGroup.getFieldRepetitionCount(0); // Get the number of elements in the vector
-      double[] vector = new double[vectorSize];
+      // Extract the vector (double[]) from the record
+      Group vectorGroup = record.getGroup("vector", 0);// Access the 'vector' field
+      int vectorSize = vectorGroup.getFieldRepetitionCount(0);// Get the number of elements in the vector
+      float[] vector = new float[vectorSize];
+      
+      // We detect the type from the schema
+      Group firstElement = vectorGroup.getGroup(0, 0);
+      boolean isDouble = firstElement.getType().getFields().get(0).asPrimitiveType().getPrimitiveTypeName().equals(PrimitiveType.PrimitiveTypeName.DOUBLE);
+      
+      // Single-pass read with conditional cast if needed
       for (int i = 0; i < vectorSize; i++) {
-        Group listGroup = vectorGroup.getGroup(0, i); // Access the 'list' group
-        vector[i] = listGroup.getDouble("element", 0); // Get the double value from the 'element' field
+        Group listGroup = vectorGroup.getGroup(0, i);
+        vector[i] = isDouble ? (float) listGroup.getDouble("element", 0) : listGroup.getFloat("element", 0);
```

Comment:
Here, you're assuming if it's not a double, it's a float, right? But what if it's an int or some other type?

How about we check explicitly for double and float, and throw an exception if it's neither?